File: packages/lhc/src/domains/thread-view/internal/boundary.ts
// Visibility-boundary read path (Story 1). The boundary is a source event
// order shared with the compact point's coordinate system; tool results
// at-or-behind it render short. Writes — the post-commit advance and
// compact's reset — land in Stories 4 and 2; this module reads only.
import type { DatabaseSync } from "node:sqlite";

// The singleton row exists from migration v6 on (seeded at position 0,
// everything full). A missing row is a damaged thread file, surfaced as a
// throw for the operation boundary's storage_failure wrap.
export function readBoundaryPosition(db: DatabaseSync): number {
  const row = db
    .prepare(`SELECT position FROM view_boundary WHERE thread_singleton = 1`)
    .get() as { position: number | bigint } | undefined;
  if (row === undefined) {
    throw new Error("view_boundary singleton row missing (migration v6 seeds it)");
  }
  return Number(row.position);
}

// The visibility zone's token sum: live (deleted-filtered) tool results ahead
// of both the boundary position and the compact point, one indexed query.
// Parameterized by position on purpose (story Technical Notes): Story 4's
// advance consumes this same function instead of re-deriving the sum, so
// "the advance's sum equals what status reports" is structural — the design
// requires the two sums equal, and sharing the query makes that a fact of
// the code rather than a discipline.
export function visibilityZoneTokens(
  db: DatabaseSync,
  position: number,
  compactPoint: number,
): number {
  const row = db
    .prepare(
      `SELECT COALESCE(SUM(token_estimate), 0) AS zone FROM message
       WHERE kind = 'tool_result' AND deleted_at IS NULL
         AND source_event_order > ? AND source_event_order > ?`,
    )
    .get(position, compactPoint) as { zone: number | bigint };
  return Number(row.zone);
}

File: packages/lhc-py/src/lhc/thread_view/internal/boundary.py
"""Ported from packages/lhc/src/thread-view/internal/boundary.ts. Phase 1 skeleton.

Visibility boundary. The boundary is a source event order shared with the
compact point's coordinate system; tool results at-or-behind it render short.
Compact resets it inside compact's own transaction.
"""

from __future__ import annotations

from ...shared_tech.storage import Database

# The singleton row is seeded at thread creation (position 0, everything full).
# A missing row is a damaged thread file, surfaced as a throw for the
# operation boundary's storage_failure wrap.
_SQL_READ_BOUNDARY_POSITION = """SELECT position FROM view_boundary WHERE thread_singleton = 1"""

_SQL_VISIBILITY_ZONE_TOKENS = (
    """SELECT COALESCE(SUM(token_estimate), 0) AS zone FROM message
       WHERE kind = 'tool_result' AND deleted_at IS NULL
         AND source_event_order > ? AND source_event_order > ?"""
)


def read_boundary_position(db: Database) -> int:
    row = db.prepare(_SQL_READ_BOUNDARY_POSITION).get()
    if row is None:
        raise RuntimeError(
            "view_boundary singleton row missing (thread creation seeds it)"
        )
    return int(row["position"])


# The visibility zone's token sum: live (deleted-filtered) tool results ahead
# of both the boundary position and the compact point, one indexed query.
def visibility_zone_tokens(db: Database, position: int, compact_point: int) -> int:
    row = db.prepare(_SQL_VISIBILITY_ZONE_TOKENS).get(position, compact_point)
    # COALESCE always yields a row under node:sqlite / our adapter.
    return int(row["zone"]) if row is not None else 0
